quadmath_format_valid rejects "%+.17g"

**Description**
PDL's recently-added use of `%+.17g` (see https://github.com/PDLPorters/pdl/blob/dc5a0284c8d70dc03bf63506edb4c6b8b9b63977/Basic/Core/pdl.h.PL#L55) is rejected as invalid on Perl 5.33.7 with quadmath. This is especially regrettable since PDL doesn't yet fully support quadmath and long doubles, it's on the roadmap to do so, and this is a bit of a blocker.

**Steps to Reproduce**
Build and test PDL 2.029 on Perl 5.33.7 with quadmath on at least Cygwin - see http://www.cpantesters.org/cpan/report/23fa1dc3-6f60-1014-9102-2ba2af1defc7

**Expected behavior**
Since https://gcc.gnu.org/onlinedocs/gcc-9.3.0/libquadmath/quadmath_005fsnprintf.html has example code with `%+`, it seems the Perl function should not reject it out of hand.

**Perl configuration**
From the CPAN Testers report above (not my system):

```
Summary of my perl5 (revision 5 version 33 subversion 7) configuration:
   
  Platform:
    osname=cygwin
    osvers=3.1.7(0.34053)
    archname=x86_64-cygwin-thread-multi-quadmath
    uname='cygwin_nt-10.0 os-a01-l0123 3.1.7(0.34053) 2020-08-22 17:48 x86_64 cygwin '
    config_args='-de -Dmksymlinks -Dusedevel -Uinc_version_list -Uinstallusrbinperl -Dusethreads -Dusemultiplicity -Dusequadmath -Dprefix=/opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath -Darchname=x86_64-cygwin-thread-multi-quadmath'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags ='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_FORTIFY_SOURCE=2'
    optimize='-O3'
    cppflags='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion=''
    gccversion='7.4.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='__float128'
    nvsize=16
    Off_t='off_t'
    lseeksize=8
    alignbytes=16
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags =' -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base'
    libpth=/cygdrive/c/share/apps/gcc-7.4.0/cygwin-3-x86_64/bin/../lib/gcc/x86_64-pc-cygwin/7.4.0/include-fixed /cygdrive/c/share/apps/gcc-7.4.0/cygwin-3-x86_64/bin/../lib/gcc/../../lib /usr/lib /usr/lib/../lib/../include/w32api /lib
    libs=-lpthread -ldl -lquadmath
    perllibs=-lpthread -ldl -lquadmath
    libc=/usr/lib/libc.a
    so=dll
    useshrplib=true
    libperl=cygperl5_33_7.dll
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags=' --shared  -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_IMPLICIT_CONTEXT
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_QUADMATH
    USE_REENTRANT_API
    USE_THREAD_SAFE_LOCALE
  Built under cygwin
  Compiled at Feb 25 2021 10:17:37
  %ENV:
    PERL5LIB="/opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/site_perl:/opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/perl5:/opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib"
    PERL5OPT=""
    PERL5_CPANPLUS_IS_RUNNING="29818"
    PERL5_CPAN_IS_RUNNING="29818"
    PERL_CPAN_REPORTER_CONFIG="/cygdrive/c/peter/.cpanreporter/config-cygwin-OS-A01-L0123.ini"
    PERL_EXTUTILS_AUTOINSTALL="1"
    PERL_MM_USE_DEFAULT="1"
    PERL_TEST_CRITIC="1"
    PERL_TEST_POD="1"
    PERL_USE_UNSAFE_INC="1"
  @INC:
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/site_perl/5.33.7/x86_64-cygwin-thread-multi-quadmath
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/site_perl/5.33.7
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/site_perl
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/perl5/5.33.7
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/perl5/x86_64-cygwin-thread-multi-quadmath
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/perl5
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/5.33.7/x86_64-cygwin-thread-multi-quadmath
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib/5.33.7
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/local/lib
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/lib/site_perl/5.33.7/x86_64-cygwin-thread-multi-quadmath
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/lib/site_perl/5.33.7
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/lib/5.33.7/x86_64-cygwin-thread-multi-quadmath
    /opt/perl5/cygwin_nt-10.0/3.1.7/x86_64/perl/5.33.7/x86_64-cygwin-thread-multi-quadmath/lib/5.33.7
    .
```

The troublesome code in pdl.h is:
```
sv_setpvf(outsv, "%.17g%+.17gi",creal(inany.value.C),cimag(inany.value.C));
```

I was of the opinion that the last 2 args must be NVs, whereas you've stated they are doubles.

So, I wrote the following script
```
use strict;
use warnings;

use Inline C => Config =>
  BUILD_NOISY => 1,
;

use Inline C => <<'EOC';

SV * foo() {
  SV *sv ;
  sv = NEWSV(0, 19);
  sv_setpvf(sv, "%+.17g", sqrt(2.0));
  return sv;
}

EOC

my $x = foo();
print "sqrt(2) is $x\n";
```
That should, of course, then run fine on a quadmath perl because the final arg supplied to sv_setpvf() is a double ... but it doesn't ... it produces the very same panic:
```
panic: quadmath invalid format "%+.17g" at try.pl line 21.
```
And, IMO, that's a bug in quadmath perl. (Surely, you should be able to assign a 'double' to an SV, even when nvtype is __float128 ?)

The preceding line in pdl.h contains:
```
sv_setpvf(outsv, "%.9g%+.9gi", crealf(inany.value.G),cimagf(inany.value.G));
```
From my testing, if that line ever gets executed on a quadmath perl, then the same panic will also be seen.
Is that so ?

--- a/ext/XS-APItest/APItest.pm
+++ b/ext/XS-APItest/APItest.pm
@@ -5,7 +5,7 @@ use strict;
 use warnings;
 use Carp;
 
-our $VERSION = '1.15';
+our $VERSION = '1.16';
 
 require XSLoader;
 

--- a/ext/XS-APItest/APItest.xs
+++ b/ext/XS-APItest/APItest.xs
@@ -2356,6 +2356,18 @@ print_long_double()
 #endif
 
 void
+print_long_doubleL()
+        CODE:
+#ifdef HAS_LONG_DOUBLE
+        /* used to test we allow the length modifier required by the standard */
+        long double val = 7.0;
+        printf("%5.3Lf\n",val);
+#else
+        double val = 7.0;
+        printf("%5.3f\n",val);
+#endif
+
+void
 print_int(val)
         int val
         CODE:

--- a/ext/XS-APItest/t/printf.t
+++ b/ext/XS-APItest/t/printf.t
@@ -1,12 +1,4 @@
-BEGIN {
-  require Config; import Config;
-  if ($Config{usequadmath}) {
-     print "1..0 # Skip: usequadmath\n";
-     exit(0);
-  }
-}
-
-use Test::More tests => 12;
+use Test::More tests => 13;
 
 BEGIN { use_ok('XS::APItest') };
 
@@ -29,6 +21,7 @@ print_int(3);
 print_long(4);
 print_float(4);
 print_long_double() if $ldok;  # val=7 hardwired
+print_long_doubleL() if $ldok;  # val=7 hardwired
 
 print_flush();
 
@@ -47,8 +40,9 @@ is($output[2], "4", "print_long");
 is($output[3], "4.000", "print_float");
 
 SKIP: {
-   skip "No long doubles", 1 unless $ldok;
+   skip "No long doubles", 2 unless $ldok;
    is($output[4], "7.000", "print_long_double");
+   is($output[5], "7.000", "print_long_doubleL");
 }
 
 {

--- a/sv.c
+++ b/sv.c
@@ -12288,10 +12288,6 @@ Perl_sv_vcatpvfn_flags(pTHX_ SV *const sv, const char *const pat, const STRLEN p
     (IVSIZE == 4 && !defined(HAS_LONG_DOUBLE))
 	case 'L':			/* Ld */
 	    /* FALLTHROUGH */
-#  ifdef USE_QUADMATH
-        case 'Q':
-	    /* FALLTHROUGH */
-#  endif
 #  if IVSIZE >= 8
 	case 'q':			/* qd */
 #  endif
@@ -12319,6 +12315,9 @@ Perl_sv_vcatpvfn_flags(pTHX_ SV *const sv, const char *const pat, const STRLEN p
 	    else
 		intsize = 'h';
 	    break;
+#ifdef USE_QUADMATH
+        case 'Q':
+#endif
 	case 'V':
 	case 'z':
 	case 't':
@@ -12840,6 +12839,10 @@ Perl_sv_vcatpvfn_flags(pTHX_ SV *const sv, const char *const pat, const STRLEN p
 	       for simplicity we allow any of %Lf, %llf, %qf for long double
 	    */
 	    switch (intsize) {
+#if defined(USE_QUADMATH)
+            case 'Q':
+                break;
+#endif
 	    case 'V':
 #if defined(USE_LONG_DOUBLE) || defined(USE_QUADMATH)
 		intsize = 'q';
@@ -12879,9 +12882,10 @@ Perl_sv_vcatpvfn_flags(pTHX_ SV *const sv, const char *const pat, const STRLEN p
                  * is when the format specifier explicitly asks so with
                  * e.g. "%Lg". */
 #ifdef USE_QUADMATH
-                fv = intsize == 'q' ?
-                    va_arg(*args, NV) : va_arg(*args, double);
-                nv = fv;
+                nv = intsize == 'Q' ? va_arg(*args, NV) :
+                    intsize == 'q' ? va_arg(*args, long double) :
+                    va_arg(*args, double);
+                fv = nv;
 #elif LONG_DOUBLESIZE > DOUBLESIZE
                 if (intsize == 'q') {
                     fv = va_arg(*args, long double);
@@ -13141,10 +13145,12 @@ Perl_sv_vcatpvfn_flags(pTHX_ SV *const sv, const char *const pat, const STRLEN p
                 *--ptr = '\0';
                 *--ptr = c;
 #if defined(USE_QUADMATH)
-		if (intsize == 'q') {
-                    /* "g" -> "Qg" */
-                    *--ptr = 'Q';
-                }
+                /* always use Q here.  my_snprint() throws an exception if we
+                   fallthrough to the double/long double code, even when the
+                   format is correct, presumably to avoid any accidentally
+                   missing Q.
+                */
+                *--ptr = 'Q';
                 /* FIXME: what to do if HAS_LONG_DOUBLE but not PERL_PRIfldbl? */
 #elif defined(HAS_LONG_DOUBLE) && defined(PERL_PRIfldbl)
 		/* Note that this is HAS_LONG_DOUBLE and PERL_PRIfldbl,

